marlowe-scripts/src/Language/Marlowe/Scripts/OpenRole.hs

```diff
@@ -0,0 +1,263 @@
+import Plutus.V1.Ledger.Value (adaSymbol, getValue, valueOf)
+import Plutus.V2.Ledger.Api (
+  Redeemer (..),
+  ScriptPurpose (Spending),
+  SerializedScript,
+  TxInInfo (TxInInfo, txInInfoOutRef, txInInfoResolved),
+  TxOut (..),
+  ValidatorHash,
+  fromBuiltinData,
+  getValidator,
+ )
+import Plutus.V2.Ledger.Api qualified as PV2
+import PlutusTx qualified
+import PlutusTx.AssocMap qualified as AssocMap
+import PlutusTx.Prelude as PlutusTxPrelude hiding (traceError, traceIfFalse)
+import Prelude qualified as Haskell
+
+#ifdef TRACE_PLUTUS
+
+import PlutusTx.Prelude (traceError, traceIfFalse)
+
+#else
+
+{-# INLINABLE traceError #-}
+traceError :: BuiltinString -> a
+traceError _ = error ()
+
+{-# INLINABLE traceIfFalse #-}
+traceIfFalse :: BuiltinString -> a -> a
+traceIfFalse _ = id
+
+#endif
+
+-- By decoding only the part of the script context I was able
+-- to bring down the size of the validator from 4928 to 4540 bytes.
+data SubScriptContext = SubScriptContext
+  { scriptContextTxInfo :: SubTxInfo
+  , scriptContextPurpose :: ScriptPurpose
+  }
+  deriving stock (Generic, Haskell.Eq, Haskell.Show)
+
+instance Eq SubScriptContext where
+  {-# INLINEABLE (==) #-}
+  SubScriptContext info purpose == SubScriptContext info' purpose' = info == info' && purpose == purpose'
+
+data SubTxInfo = SubTxInfo
+  { txInfoInputs :: [TxInInfo]
+  , txInfoReferenceInputs :: BuiltinData
+  , txInfoOutputs :: BuiltinData
+  , txInfoFee :: BuiltinData
+  , txInfoMint :: BuiltinData
+  , txInfoDCert :: BuiltinData
+  , txInfoWdrl :: BuiltinData
+  , txInfoValidRange :: BuiltinData
+  , txInfoSignatories :: BuiltinData
+  , txInfoRedeemers :: AssocMap.Map ScriptPurpose Redeemer
+  , txInfoData :: BuiltinData
+  , txInfoId :: BuiltinData
+  }
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+
+instance Eq SubTxInfo where
+  {-# INLINEABLE (==) #-}
+  SubTxInfo i ri o f m c w r s rs d tid == SubTxInfo i' ri' o' f' m' c' w' r' s' rs' d' tid' =
+    i
+      == i'
+      && ri
+      == ri'
+      && o
+      == o'
+      && f
+      == f'
+      && m
+      == m'
+      && c
+      == c'
+      && w
+      == w'
+      && r
+      == r'
+      && s
+      == s'
+      && rs
+      == rs'
+      && d
+      == d'
+      && tid
+      == tid'
+
+{- Open Role validator - it releases role token(s) (you can put more coins of the same role token to it) based on a few conditions:
+
+   1. Value should contain only min. ada and a specific role token(s).
+   2. Transaction should spend Marlowe output which contains corresponding thread token (the same currency symbol as
+   role and token name as defined in the inlined datum).
+   3. The list of Marlowe inputs should contain at least one action either `IDeposit` or `IChoice` dedicated for the same role
+   party as the locked role token.
+
+   In the current version we assume that thread token has the same currency as the role token which can be actually undesired.
```

Comment:
```suggestion
   In the current version we assume that thread token has the same currency as the role token, even though this might actually be undesired.
```